src/Npgsql/NpgsqlConnection.cs

```diff
@@ -1806,9 +1807,12 @@ public NpgsqlConnection CloneWith(string connectionString)
         if (csb.PersistSecurityInfo && !Settings.PersistSecurityInfo)
             csb.PersistSecurityInfo = false;
 
-        return new NpgsqlConnection(csb.ToString()) {
-            ProvideClientCertificatesCallback = ProvideClientCertificatesCallback,
-            UserCertificateValidationCallback = UserCertificateValidationCallback,
+        return new NpgsqlConnection(csb.ToString())
+        {
+            ProvideClientCertificatesCallback = ProvideClientCertificatesCallback ?? (_dataSource?.ClientCertificatesCallback is null
+                ? null
+                : (ProvideClientCertificatesCallback)(certs => _dataSource?.ClientCertificatesCallback(certs))),
```

Comment:
That's probably not a good idea because there is a closure on `_dataSource` (which can change)...

Reply:
Good catch...